### PR TITLE
[#2021] move the rest-client from cli module to common module

### PR DESCRIPTION
**Pull request by xianjingfeng**
### What changes were proposed in this pull request?

1. Move the rest-client from cli module to common module.
2. Some other small optimizations.

### Why are the changes needed?

Fix: #2021 

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?

CI

### CODE REVIEW

**Comment by maobaolong** [comment]
@xianjingfeng I suppose to abstract a new module `server-common` by  https://github.com/apache/incubator-uniffle/pull/1884 aimed to reduce the client fat jar size.

If we move rest-client from cli to common, the uniffle client fat jar would be greater than before although the shaded rss spark client never use the rest-client?

So we can move the `rest-client` to `server-common`, and all server-side modules can use `server-common` as dependency, and rss-client can avoid introduce the extra unnecessary dependencies.

**Comment by xianjingfeng** [comment]
> @xianjingfeng I suppose to abstract a new module `server-common` by #1884 aimed to reduce the client fat jar size.
> 
> If we move rest-client from cli to common, the uniffle client fat jar would be greater than before although the shaded rss spark client never use the rest-client?
> 
> So we can move the `rest-client` to `server-common`, and all server-side modules can use `server-common` as dependency, and rss-client can avoid introduce the extra unnecessary dependencies.

Agree. But when will `#1884` be ready?

File: common/pom.xml [hunk 1]
```diff
@@ -174,6 +174,14 @@
       <groupId>org.apache.hbase.thirdparty</groupId>
       <artifactId>hbase-shaded-jersey</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.projectlombok</groupId>
+      <artifactId>lombok</artifactId>
```
[the remark this answers is not in the document]
Comment (xianjingfeng) [reply common/pom.xml:179]: Why? It is provided.

**Comment by xianjingfeng** [comment]
> One high level question: why does the coordinator module need the rest client?

For #1941. The shuffle server will provide a http interface for obtaining application information. The coordinator needs to obtain this data from each shuffle servers and then aggregate them.

**Comment by xianjingfeng** [comment]
> Why RPC calls cannot be used? I think it will more efficient and doesn't rely on the http interface.

 We need obtain this data from a single shuffle server sometimes, so we need provide a http interface in the shuffle server side. I think it is not necessary provide another RPC interface. The call frequency of this interface is not high.

**Comment by xianjingfeng** [comment]
> Like that the shuffle servers could report all its running applications back to coordinator via rpc calls? Just like how the heartbeat is made.

The amount of the data is a bit large. The data of an application would be as follow.

```java
public class ApplicationMetricsVo {
  private long totalDataSize;
  private long inMemoryDataSize;
  private long onLocalFileDataSize;
  private long onHadoopDataSize;
  private Map<Integer, Map<Integer, Long>> partitionDataSizes;
  private boolean existHugePartition;
}
```

File: common/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (xianjingfeng) [reply common/pom.xml:179]: Ok for me to remove it.

**Comment by maobaolong** [comment]
> > @xianjingfeng I suppose to abstract a new module `server-common` by #1884 aimed to reduce the client fat jar size.
> > If we move rest-client from cli to common, the uniffle client fat jar would be greater than before although the shaded rss spark client never use the rest-client?
> > So we can move the `rest-client` to `server-common`, and all server-side modules can use `server-common` as dependency, and rss-client can avoid introduce the extra unnecessary dependencies.
> 
> Agree. But when will #1884 be ready?

@xianjingfeng I just refactor my PR and keep only create a server-common module, please take a look.

**Comment by xianjingfeng** [comment]
> > > Like that the shuffle servers could report all its running applications back to coordinator via rpc calls? Just like how the heartbeat is made.
> > 
> > The amount of the data is a bit large. The data of an application would be as follow.
> > 
> > ```java
> > public class ApplicationMetricsVo {
> >   private long totalDataSize;
> >   private long inMemoryDataSize;
> >   private long onLocalFileDataSize;
> >   private long onHadoopDataSize;
> >   private Map<Integer, Map<Integer, Long>> partitionDataSizes;
> >   private boolean existHugePartition;
> > }
> > ```
> 
> Hmmm, looks pretty compact to me. Except the partition data sizes, I don't think this info has much different than the heartbeat message. 
> 
> If you haven't finished all the related code, I would suggest to implement a new `ReportRunningApplications` service point in coordinator instead, which I think would be less code.
> 
> The shuffle server could still export its internal metrics/running applications via the Web interface.

If this method is used, each shuffle server needs to report hundreds of application information and tens of thousands of partition information.This may cause performance issues for the coordinator.

In general, we only need to check the information of a particular application, not all of them.

**Comment by xianjingfeng** [comment]
> > In general, we only need to check the information of a particular application, not all of them.
> 
> Could you elaborate a bit more on this? Based on your umbrella issue(#1941), I think we still need all the application infos to detect abnormal apps.

Scenario 1:
When we find that the cluster load is high, we need to know whether the shuffle data volume of some applications is too large. In this scenario, we need the brief information about all applications, such as the total shuffle data volume.

Scenario 2:
When we find that an application is very slow, we need to quickly determine whether the application has data skew. In this scenario, we need to know the detailed information of this application, including the amount of data in each partition.

In general, we only query this data when something unusual happens, and there is no need to report the data to the coordinator regularly.

**Comment by xianjingfeng** [comment]
> I see, it's kind of throw away messages and for diagnostic purpose. I'm neutral to this change then. However I still think it would be better to report application info back and that can cover all the use cases you have described. The additional overhead might be negligible compared to other infos.
> 
> > Scenario 1: When we find that the cluster load is high, we need to know whether the shuffle data volume of some applications is too large. In this scenario, we need the brief information about all applications, such as the total shuffle data volume.
> > Scenario 2: When we find that an application is very slow, we need to quickly determine whether the application has data skew. In this scenario, we need to know the detailed information of this application, including the amount of data in each partition.
> > In general, we only query this data when something unusual happens, and there is no need to report the data to the coordinator regularly.

What's your opinions? @jerqi @zuston @rickyma @maobaolong

File: common/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (maobaolong) [reply common/pom.xml:179]: @advancedxy Currently, the spark-client could start a GrpcServer for ShuffleManagerGrpcService, so nearly all the module dependent web\metrics\prometheus.

We should find a way to do some refactor to reduce the dependencies of client jar, then we can reduce class conflict and get a tiny size client jar.

This is not related to this PR.

**Comment by maobaolong** [comment]
@xianjingfeng Thanks for invite me to join this discussion. In fact we have done some work on displaying the application detailed information in coordinator. 

- Application buffer(size/count) flush(size/filecount)
- Application exit reason

We also want to show the application `spark-conf` and find some lineage from the spark application and the upstream scheduled task, base on this, we can get effect and benefit after a spark job using Uniffle.

We also want to show more detailed include each stage write/read throughput cost and timeline. 

All of these, we want to store and persist so that we can search the history information, maybe this can help us on troubleshooting.

**Comment by xianjingfeng** [comment]
> After all of above, we want to store and persist into storage, so that we can search the history information, maybe this can help us in daily troubleshooting.

When will this be done?


> Do you think we have to know each partition information? If we only report `topN` partition information to coordinator, do you think it can be enough and easy? If so, we can add some extra information in the heartbeat request of Driver->Coordinator and Server->Coordinator, the Coordinator can have a detailed Application information.

Ok for me. And when will this be done?

**Comment by maobaolong** [comment]
@lwllvyb Would you like to show some progress about the work about collect and show application detailed in coordinator and dashboard?

**Comment by maobaolong** [comment]
> > @xianjingfeng I suppose to abstract a new module `server-common` by #1884 aimed to reduce the client fat jar size.
> > If we move rest-client from cli to common, the uniffle client fat jar would be greater than before although the shaded rss spark client never use the rest-client?
> > So we can move the `rest-client` to `server-common`, and all server-side modules can use `server-common` as dependency, and rss-client can avoid introduce the extra unnecessary dependencies.
> 
> Agree. But when will #1884 be ready?

@xianjingfeng Glad to share that the #1884 has been merged.

**Comment by xianjingfeng** [comment]
> @xianjingfeng Glad to share that the #1884 has been merged.

Get it

**Comment by xianjingfeng** [comment]
This PR is no longer needed. Let's wait for @lwllvyb

**Comment by lwllvyb** [comment]
> Recntly, I was working on how to display more information on the dashboard UI. More details about the application will be displayed in the Application Page.（I am new to Vue.js, can somebody help beautify the UI?）
> 
> I add some columns in the Application Page: PartitionNum | writeMemorySize | flushedLocalFileNum | flushedLocalTotalSize | flushedHadoopFileNum | flushedHadoopTotalSize | writeTotalSize
> 
> image

I will create a new PR.

**Pull request closed without merging**